File: langs/javascript/tests/workspace.test.ts
import { describe, it, expect } from "vitest";
import { findMonorepoRoot, getSharedEnv, generateTurboEnvConfig } from "../src/workspace.js";
import { mkdtempSync, writeFileSync, mkdirSync, rmSync } from "node:fs";
import { join } from "node:path";
import { tmpdir } from "node:os";

describe("Workspace / Monorepo Support", () => {
  describe("findMonorepoRoot", () => {
    it("detects pnpm workspace", () => {
      const tmp = mkdtempSync(join(tmpdir(), "dotlyte-ws-"));
      writeFileSync(join(tmp, "pnpm-workspace.yaml"), "packages:\n  - 'apps/*'\n  - 'packages/*'\n");
      const child = join(tmp, "apps", "web");
      mkdirSync(child, { recursive: true });

      const info = findMonorepoRoot(child);
      expect(info).toBeDefined();
      expect(info!.root).toBe(tmp);
      expect(info!.type).toBe("pnpm");
      expect(info!.packages).toContain("apps/*");

      rmSync(tmp, { recursive: true, force: true });
    });

    it("detects npm workspace from package.json", () => {
      const tmp = mkdtempSync(join(tmpdir(), "dotlyte-ws-"));
      writeFileSync(
        join(tmp, "package.json"),
        JSON.stringify({ name: "monorepo", workspaces: ["packages/*"] }),
      );
      const child = join(tmp, "packages", "lib");
      mkdirSync(child, { recursive: true });

      const info = findMonorepoRoot(child);
      expect(info).toBeDefined();
      expect(info!.type).toBe("npm");
      expect(info!.packages).toContain("packages/*");

      rmSync(tmp, { recursive: true, force: true });
    });

    it("detects turbo.json monorepo", () => {
      const tmp = mkdtempSync(join(tmpdir(), "dotlyte-ws-"));
      writeFileSync(join(tmp, "turbo.json"), JSON.stringify({ pipeline: {} }));
      writeFileSync(
        join(tmp, "package.json"),
        JSON.stringify({ workspaces: ["apps/*"] }),
      );
      const child = join(tmp, "apps", "web");
      mkdirSync(child, { recursive: true });

      const info = findMonorepoRoot(child);
      expect(info).toBeDefined();
      expect(info!.type).toBe("turbo");

      rmSync(tmp, { recursive: true, force: true });
    });

    it("returns undefined when not in a monorepo", () => {
      const tmp = mkdtempSync(join(tmpdir(), "dotlyte-no-ws-"));
      const info = findMonorepoRoot(tmp);
      // May detect parent dirs, so just check it doesn't crash
      expect(info === undefined || info.root !== tmp).toBe(true);
      rmSync(tmp, { recursive: true, force: true });
    });
  });

  describe("getSharedEnv", () => {
    it("reads .env from root and returns key-value pairs", () => {
      const tmp = mkdtempSync(join(tmpdir(), "dotlyte-se-"));
      writeFileSync(join(tmp, ".env"), "DB_HOST=localhost\nPORT=3000\n# comment\n");

      const result = getSharedEnv(tmp);
      expect(result.DB_HOST).toBe("localhost");
      expect(result.PORT).toBe("3000");

      rmSync(tmp, { recursive: true, force: true });
    });

    it("strips prefix from keys", () => {
      const tmp = mkdtempSync(join(tmpdir(), "dotlyte-se-"));
      writeFileSync(join(tmp, ".env"), "APP_DB_HOST=localhost\nAPP_PORT=3000\n");

      const result = getSharedEnv(tmp, "APP_");
      expect(result.DB_HOST).toBe("localhost");
      expect(result.PORT).toBe("3000");

      rmSync(tmp, { recursive: true, force: true });
    });

    it("returns empty object when .env missing", () => {
      const tmp = mkdtempSync(join(tmpdir(), "dotlyte-se-"));
      const result = getSharedEnv(tmp);
      expect(result).toEqual({});
      rmSync(tmp, { recursive: true, force: true });
    });
  });

  describe("generateTurboEnvConfig", () => {
    it("generates globalEnv config", () => {
      const result = generateTurboEnvConfig(
        ["DB_HOST", "PORT"],
        { global: true },
      );
      expect(result).toEqual({ globalEnv: ["DB_HOST", "PORT"] });
    });

    it("generates env config (non-global)", () => {
      const result = generateTurboEnvConfig(["API_URL"]);
      expect(result).toEqual({ env: ["API_URL"] });
    });
  });
});

File: langs/rust/tests/workspace_test.rs
//! Integration tests for the workspace/monorepo system.

use std::fs;

use dotlyte::workspace::{find_monorepo_root, get_shared_env};

#[test]
fn test_find_monorepo_root_from_nonexistent_dir() {
    // Should fail gracefully for a directory that doesn't exist
    let result = find_monorepo_root(Some("/tmp/dotlyte_nonexistent_test_dir_xyz"));
    assert!(result.is_err());
}

#[test]
fn test_find_monorepo_root_detects_pnpm() {
    let dir = tempfile::tempdir().unwrap();
    let ws_file = dir.path().join("pnpm-workspace.yaml");
    fs::write(
        &ws_file,
        "packages:\n  - 'packages/core'\n  - 'packages/cli'\n",
    )
    .unwrap();

    let result = find_monorepo_root(Some(dir.path().to_str().unwrap()));
    assert!(result.is_ok());
    let info = result.unwrap();
    assert_eq!(info.monorepo_type, "pnpm");
}

#[test]
fn test_find_monorepo_root_detects_turbo() {
    let dir = tempfile::tempdir().unwrap();
    fs::write(dir.path().join("turbo.json"), "{}").unwrap();
    fs::write(dir.path().join("package.json"), r#"{"workspaces":["apps/*"]}"#).unwrap();

    let result = find_monorepo_root(Some(dir.path().to_str().unwrap()));
    assert!(result.is_ok());
    let info = result.unwrap();
    assert_eq!(info.monorepo_type, "turbo");
}

#[test]
fn test_find_monorepo_root_detects_go_work() {
    let dir = tempfile::tempdir().unwrap();
    fs::write(
        dir.path().join("go.work"),
        "go 1.21\n\nuse (\n\t./cmd\n\t./pkg\n)\n",
    )
    .unwrap();

    let result = find_monorepo_root(Some(dir.path().to_str().unwrap()));
    assert!(result.is_ok());
    let info = result.unwrap();
    assert_eq!(info.monorepo_type, "go");
}

#[test]
fn test_find_monorepo_root_detects_npm_workspaces() {
    let dir = tempfile::tempdir().unwrap();
    fs::write(
        dir.path().join("package.json"),
        r#"{"workspaces": ["packages/*"]}"#,
    )
    .unwrap();

    let result = find_monorepo_root(Some(dir.path().to_str().unwrap()));
    assert!(result.is_ok());
    let info = result.unwrap();
    assert_eq!(info.monorepo_type, "npm");
}

#[test]
fn test_find_monorepo_root_detects_yarn() {
    let dir = tempfile::tempdir().unwrap();
    fs::write(
        dir.path().join("package.json"),
        r#"{"workspaces": ["packages/*"]}"#,
    )
    .unwrap();
    fs::write(dir.path().join("yarn.lock"), "").unwrap();

    let result = find_monorepo_root(Some(dir.path().to_str().unwrap()));
    assert!(result.is_ok());
    let info = result.unwrap();
    assert_eq!(info.monorepo_type, "yarn");
}

#[test]
fn test_get_shared_env_no_file() {
    let dir = tempfile::tempdir().unwrap();
    let result = get_shared_env(dir.path().to_str().unwrap(), None).unwrap();
    assert!(result.is_empty());
}

#[test]
fn test_get_shared_env_basic() {
    let dir = tempfile::tempdir().unwrap();
    fs::write(dir.path().join(".env"), "PORT=3000\nHOST=localhost\n").unwrap();

    let result = get_shared_env(dir.path().to_str().unwrap(), None).unwrap();
    assert!(!result.is_empty());
    assert!(result.contains_key("PORT"));
    assert!(result.contains_key("HOST"));
}

#[test]
fn test_get_shared_env_with_prefix() {
    let dir = tempfile::tempdir().unwrap();
    fs::write(
        dir.path().join(".env"),
        "APP_DB_HOST=localhost\nAPP_DB_PORT=5432\nOTHER=value\n",
    )
    .unwrap();

    let result = get_shared_env(dir.path().to_str().unwrap(), Some("APP")).unwrap();
    // Should contain stripped keys converted to dot notation
    assert!(result.contains_key("db.host"));
    assert!(result.contains_key("db.port"));
    // Should NOT contain non-matching keys
    assert!(!result.contains_key("OTHER"));
    assert!(!result.contains_key("other"));
}
